In [0]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline

In [0]:
# Importing the necessary libraries!
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
# Importing the necessary libraries!
from sklearn.model_selection import train_test_split
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout
import numpy as np
from keras import Sequential
from keras.layers import Dense

Using TensorFlow backend.


# WHITE WINE - PREDICTION OF QUALITY USING NEURAL NETWORKS



In [0]:
# Loading Dataset to a pandas dataframe
df_white = pd.read_csv(
    'http://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-white.csv', delimiter=";")

In [25]:
# Checking if the data has loaded properly!
df_white.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [26]:
 print('shape of the dataset:', df_white.shape)

shape of the dataset: (4898, 12)


In [27]:
df_white.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
fixed acidity           4898 non-null float64
volatile acidity        4898 non-null float64
citric acid             4898 non-null float64
residual sugar          4898 non-null float64
chlorides               4898 non-null float64
free sulfur dioxide     4898 non-null float64
total sulfur dioxide    4898 non-null float64
density                 4898 non-null float64
pH                      4898 non-null float64
sulphates               4898 non-null float64
alcohol                 4898 non-null float64
quality                 4898 non-null int64
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


In [28]:
# Renaming the column headers without spaces in the names.
df_white.rename(columns={'fixed acidity': 'fixed_acidity','citric acid':'citric_acid',
                         'volatile acidity':'volatile_acidity','residual sugar':'residual_sugar',
                         'free sulfur dioxide':'free_sulfur_dioxide','total sulfur dioxide':'total_sulfur_dioxide'}, inplace=True)
df_white.head(n=5)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [29]:
df_white.quality.unique()

array([6, 5, 7, 8, 4, 3, 9])

**The quality ranges from 1 to 10 - 1 being poor and 10 being the best. We can see the dataset has values 3,4,5,6,7,8,9. We can see we do not have any observations for 1,2 and 10 in our dataset.**

In [30]:
df_white.quality.value_counts().sort_index()

3      20
4     163
5    1457
6    2198
7     880
8     175
9       5
Name: quality, dtype: int64

**We can see the dataset has values 3,4,5,6,7,8,9. We can see we do not have any observations for 1,2 and 10 in our dataset. 


In [0]:
np.random.seed(3)

In [0]:
# Number of classes for quality ( t to 10 - 1 being Poor 10 being Best)
classes = 10

In [0]:
# Splitting quality into y and keeping the rest of features in X
X = df_white.iloc[:,0:11]
y = df_white.iloc[:,11]

In [34]:
X.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  
0      8.8  
1      9.5  
2     10.1  
3      9.9  
4      9.9

In [35]:
y.head()

0    6
1    6
2    6
3    6
4    6
Name: quality, dtype: int64

In [0]:
#standardizing the input feature
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)


In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [0]:
# one hot-encoding
y_train = keras.utils.to_categorical(y_train-1, classes)
y_test = keras.utils.to_categorical(y_test-1, classes)

In [0]:
classifier = Sequential()
# First Hidden Layer
classifier.add(Dense(30, activation='relu', kernel_initializer='random_normal', input_dim=11))
# Second  Hidden Layer
classifier.add(Dense(20, activation='relu', kernel_initializer='random_normal'))
# Output Layer
classifier.add(Dense(10, activation='softmax', kernel_initializer='random_normal'))

In [0]:
# Compiling the neural network
classifier.compile(optimizer ='adam',loss='categorical_crossentropy', metrics =['accuracy'])

In [45]:
# Fitting the data to the training dataset
classifier.fit(X_train,y_train, batch_size=10, epochs=250)

Epoch 1/250
3428/3428 [==============================] - 0s 104us/step - loss: 0.8757 - acc: 0.6199
Epoch 2/250
3428/3428 [==============================] - 0s 105us/step - loss: 0.8757 - acc: 0.6286
Epoch 3/250
3428/3428 [==============================] - 0s 127us/step - loss: 0.8729 - acc: 0.6243
Epoch 4/250
3428/3428 [==============================] - 0s 129us/step - loss: 0.8748 - acc: 0.6260
Epoch 5/250
3428/3428 [==============================] - 0s 132us/step - loss: 0.8713 - acc: 0.6243
Epoch 6/250
3428/3428 [==============================] - 0s 123us/step - loss: 0.8697 - acc: 0.6222
Epoch 7/250
3428/3428 [==============================] - 0s 108us/step - loss: 0.8696 - acc: 0.6208
Epoch 8/250
3428/3428 [==============================] - 0s 103us/step - loss: 0.8680 - acc: 0.6214
Epoch 9/250
3428/3428 [==============================] - 0s 104us/step - loss: 0.8647 - acc: 0.6208
Epoch 10/250
3428/3428 [==============================] - 0s 103us/step - loss: 0.8656 - acc: 0.6208

In [46]:
eval_model=classifier.evaluate(X_train, y_train)
eval_model

3428/3428 [==============================] - 0s 19us/step


[0.7432772168061558, 0.6872812135355892]

In [48]:
scores = classifier.evaluate(X_test, y_test)
print("\nAccuracy: %.2f%%" % (scores[1]*100))



1470/1470 [==============================] - 0s 21us/step

Accuracy: 53.27%


# RED WINE - PREDICTION OF QUALITY USING NEURAL NETWORKS


In [0]:
# Loading the dataset for Red wine
df_red = pd.read_csv(
    'http://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv', delimiter=";")

In [50]:
# Checking if the data has loaded properly!
df_red.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [51]:
 print('shape of the dataset:', df_red.shape)

shape of the dataset: (1599, 12)


In [52]:
df_red.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64
chlorides               1599 non-null float64
free sulfur dioxide     1599 non-null float64
total sulfur dioxide    1599 non-null float64
density                 1599 non-null float64
pH                      1599 non-null float64
sulphates               1599 non-null float64
alcohol                 1599 non-null float64
quality                 1599 non-null int64
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [53]:
# Renaming the column headers without spaces in the names.
df_red.rename(columns={'fixed acidity': 'fixed_acidity','citric acid':'citric_acid',
                         'volatile acidity':'volatile_acidity','residual sugar':'residual_sugar',
                         'free sulfur dioxide':'free_sulfur_dioxide','total sulfur dioxide':'total_sulfur_dioxide'}, inplace=True)
df_red.head(n=5)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [54]:
df_red.quality.unique()

array([5, 6, 7, 4, 8, 3])

In [55]:
df_red.quality.value_counts().sort_index()

3     10
4     53
5    681
6    638
7    199
8     18
Name: quality, dtype: int64

In [0]:
np.random.seed(3)

In [0]:
# Number of classes for quality ( t to 10 - 1 being Poor , 10 being Best)
classes = 10

In [0]:
# Splitting quality into y and keeping the rest of features in X
X = df_red.iloc[:,0:11]
y = df_red.iloc[:,11]

In [61]:
X.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  
0      9.4  
1      9.8  
2      9.8  
3      9.8  
4      9.4

In [62]:
y.head()

0    5
1    5
2    5
3    6
4    5
Name: quality, dtype: int64

In [0]:
#standardizing the input feature
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)


In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [0]:
# one hot-encoding
y_train = keras.utils.to_categorical(y_train-1, classes)
y_test = keras.utils.to_categorical(y_test-1, classes)

In [0]:
classifier = Sequential()
# First Hidden Layer
classifier.add(Dense(30, activation='relu', kernel_initializer='random_normal', input_dim=11))
# Second  Hidden Layer
classifier.add(Dense(20, activation='relu', kernel_initializer='random_normal'))
# Output Layer
classifier.add(Dense(10, activation='softmax', kernel_initializer='random_normal'))

In [0]:
# Compiling the neural network
classifier.compile(optimizer ='adam',loss='categorical_crossentropy', metrics =['accuracy'])

In [70]:
# Fitting the data to the training dataset
classifier.fit(X_train,y_train, batch_size=10, epochs=250)

Epoch 1/250
1119/1119 [==============================] - 0s 139us/step - loss: 0.7593 - acc: 0.6685
Epoch 2/250
1119/1119 [==============================] - 0s 133us/step - loss: 0.7598 - acc: 0.6792
Epoch 3/250
1119/1119 [==============================] - 0s 143us/step - loss: 0.7567 - acc: 0.6819
Epoch 4/250
1119/1119 [==============================] - 0s 133us/step - loss: 0.7542 - acc: 0.6774
Epoch 5/250
1119/1119 [==============================] - 0s 130us/step - loss: 0.7563 - acc: 0.6774
Epoch 6/250
1119/1119 [==============================] - 0s 137us/step - loss: 0.7531 - acc: 0.6872
Epoch 7/250
1119/1119 [==============================] - 0s 144us/step - loss: 0.7551 - acc: 0.6693
Epoch 8/250
1119/1119 [==============================] - 0s 139us/step - loss: 0.7487 - acc: 0.6756
Epoch 9/250
1119/1119 [==============================] - 0s 136us/step - loss: 0.7534 - acc: 0.6765
Epoch 10/250
1119/1119 [==============================] - 0s 132us/step - loss: 0.7469 - acc: 0.6702

In [71]:
eval_model=classifier.evaluate(X_train, y_train)
eval_model

1119/1119 [==============================] - 0s 82us/step


[0.4850102359690338, 0.7864164425604465]

In [72]:
scores = classifier.evaluate(X_test, y_test)
print("\nAccuracy: %.2f%%" % (scores[1]*100))



480/480 [==============================] - 0s 26us/step

Accuracy: 62.29%


# CONCLUSION - 

Accuracy for white wine dataset ( 10 classes for 250 epochs ) - 54.83%


Accuracy for red wine dataset ( 10 classes for 250 epochs ) - 62.29%

This accuracy is for the test set that has the ratings ( 3, 4, 5, 6, 7, 8, 9 for white wine ) & ( 3, 4 ,5, 6, 7, 8 for red wine ). Moreover, the network hasnt seen observations for quality ratings ( 1, 2 and 10 for white wine ) and (1, 2, 9,10 for red wine). 

I think the best way to approach would be to have three categories - Poor, Average and Best . Wine_NN.ipynb has the code for the same. 